### ISSUE TITLE
rmsprop implementation in optax and torch

### ISSUE THREAD

**Issue by vwxyzjn** [google-deepmind/optax#532]
Hello, thanks for this helpful library.

**I was wondering if `optax`'s rmsprop implementation is equivalent to torch's rmsprop implementation.** 

Recently, I have been working on a distributed reinforcement learning library called Cleanba which replicates [IMPALA](https://arxiv.org/pdf/1802.01561.pdf). I then compared our Cleanba's impala against [torchbeast](https://github.com/facebookresearch/torchbeast)'s IMPALA. I considered the following four optimizer settings:

<table>
<tr>
<td> torch setting in torchbeast </td> <td> optax setting in our cleanba</td>
</tr>
<tr>
<td>

```python
optimizer = torch.optim.Adam(
    learner_model.parameters(),
    lr=0.00025,
    eps=1e-5,
)
# ...
nn.utils.clip_grad_norm_(learner_model.parameters(), 0.5)
```

</td>
<td>
    
```python
optax.chain(
    optax.clip_by_global_norm(0.5),
    optax.adam(
        learning_rate=0.00025, 
        eps=1e-5
    ),
)
```
</td>
</tr>
</table>

<table>
<tr>
<td> torch setting in torchbeast </td> <td> optax setting in our cleanba</td>
</tr>
<tr>
<td>

```python
optimizer = torch.optim.RMSprop(
    learner_model.parameters(),
    lr=0.0006,
    momentum=0,
    eps=0.01,
    alpha=0.99,
)
# ...
nn.utils.clip_grad_norm_(learner_model.parameters(), 40)
```

</td>
<td>
    
```python
optax.chain(
    optax.clip_by_global_norm(40),
    optax.rmsprop(
         learning_rate=0.0006,
         eps=0.01,
         decay=0.99, # decay corresponds to torch's `alpha`, right?
    ),
)
```
</td>
</tr>
</table>

The results are presented in the figure below. When using the same Adam optimizer settings in torchbeast and cleanba, torchbeast's IMPALA and cleanba's IMPALA has similar performance. However, when using the same RMSprop optimizer settings, the performance differs significantly, with torchbeast's IMPALA obtaining much higher median human normalized score. This leads me to wonder if optax has the same RMSprop implementation as torch... Was wondering if you would have any thoughts on the performance discrepancies. Thanks!

<img width="1361" alt="image" src="https://github.com/deepmind/optax/assets/5555347/6915f255-c8e2-4d10-9f2e-a7efcb22d8f1">

The experiments were run for **three random seeds**, and the individual learning curves suggest torchbeast's IMPALA seems to perform qualitatively better. 

![main_10CPU](https://github.com/deepmind/optax/assets/5555347/50ad30ca-bd62-46ba-bca5-a07ee6a636c2)

**Comment by Rupt** [comment]
@vwxyzjn both of your "cleanba" examples use `optax.adam`, although the second has an invalid `decay=` argument. Is this an error is your reproduction in this issue?

**Comment by Rupt** [comment]
> This leads me to wonder if optax has the same RMSprop implementation as torch... Was wondering if you would have any thoughts on the performance discrepancies. Thanks!

@vwxyzjn Yes, the implementations differ. They treat `eps` differently.

In the denominator, optax uses $\sqrt{v + \varepsilon}$ (see [here](https://github.com/deepmind/optax/blob/master/optax/_src/alias.py#L709) and [here](https://github.com/deepmind/optax/blob/master/optax/_src/transform.py#L251)), and torch uses $(\sqrt v + \varepsilon)$ (see [here](https://pytorch.org/docs/stable/generated/torch.optim.RMSprop.html)).

**Comment by vwxyzjn** [comment]
Hi @Rupt thanks for the reply. The snippet I shared had a typo, it should be `optax.rmsprop`, which is actually used in the [experiment](https://wandb.ai/openrlbenchmark/cleanba/runs/dpaxj8ja/code?workspace=user-costa-huang)

<img width="1099" alt="image" src="https://github.com/deepmind/optax/assets/5555347/d93cd21d-f9ba-4691-bf9b-cffd9e3b8316">


Also thanks for checking the differences, so to create a `torch` equivalent optimizer, I just need to do the following? 

```diff
  def update_fn(updates, state, params=None):
    del params
    nu = update_moment_per_elem_norm(updates, state.nu, decay, 2)
    updates = jax.tree_util.tree_map(
-        lambda g, n: g * jax.lax.rsqrt(n + eps), updates, nu)
+        lambda g, n: g * jax.lax.rsqrt(n) + eps, updates, nu)
    return updates, ScaleByRmsState(nu=nu)
```

**Comment by Rupt** [comment]
Very welcome @vwxyzjn.

Not quite, the eps should still be in the denominator. I think you want this:
```diff
-        lambda g, n: g * jax.lax.rsqrt(n + eps), updates, nu)
+        lambda g, n: g / (jax.lax.sqrt(n) + eps), updates, nu)
 ```

**Comment by vwxyzjn** [comment]
Thanks so much @Rupt I will give it a try :)

**Comment by mtthss** [comment]
@vwxyzjn did this work?

**Comment by vwxyzjn** [comment]
@mtthss Actually yeah, I used the optimizer @Rupt suggested and can **now match the performance of `monobeast` with RMSprop**. The PyTorch RMSprop setting seems quite good, giving a performance boost in almost all the games tested. 

Would you like a PR to add a warning like #571?



https://github.com/vwxyzjn/cleanba/blob/d0f5edebe8539231855d657e57e46daf7c590bc7/cleanba/cleanba_impala_envpool_machado_atari_wrapper_rmsprop_pt.py#L135-L179

<img width="1285" alt="image" src="https://github.com/deepmind/optax/assets/5555347/b60a6836-36f1-4858-8632-5d73a3cf66dd">

![main_10CPU_sample_walltime_efficiency](https://github.com/deepmind/optax/assets/5555347/8ed4d55a-e2f4-4a25-a1dc-be09cd00c3e3)

![main_10CPU](https://github.com/deepmind/optax/assets/5555347/9a0ad862-6c88-4a68-b199-326a3cd1700d)

**Comment by Rupt** [comment]
@vwxyzjn Glad it worked, thanks for sharing this nice reproduction. 

> The PyTorch RMSprop setting seems quite good, giving a performance boost in almost all the games tested.

Note that the two implementations will prefer different `eps` values. You might get more similar results if using `eps**2` in the optax version, because for small nu we get $1/\sqrt{0+\varepsilon^2} = 1/(\sqrt 0 + \varepsilon)$

**Comment by mtthss** [comment]
> Would you like a PR to add a warning like https://github.com/google-deepmind/optax/issues/571?

@vwxyzjn that would be great thanks!
